Buildbot failures on latest HPX stable

## Expected Behavior

Normal build

## Actual Behavior

The build fails due to template instantiation errors.   See http://ktau.nic.uoregon.edu:8020/#/builders/7/builds/299


## Steps to Reproduce the Problem

... Please be as specific as possible while describing how to reproduce your problem.

  1. Checkout HPX stable
  1. cmake -DCMAKE_CXX_COMPILER=/packages/gcc/7.1/bin/g++ -DCMAKE_C_COMPILER=/packages/gcc/7.1/bin/gcc -DCMAKE_BUILD_TYPE=Debug -DBOOST_ROOT=/usr/local/packages/boost/1.65.0-gcc7 -DTCMALLOC_ROOT=/usr/local/packages/gperftools/git-gcc7 -DHPX_WITH_MALLOC=tcmalloc -DHWLOC_ROOT= -DCMAKE_INSTALL_PREFIX=. -DHPX_WITH_THREAD_IDLE_RATES=ON -DHPX_WITH_PARCELPORT_MPI=ON -DHPX_WITH_PARCEL_COALESCING=OFF -DHPX_WITH_TOOLS=OFF -DHPX_WITH_TESTS=ON -DHPX_WITH_EXAMPLES=OFF -DHPX_WITH_THREAD_MANAGER_IDLE_BACKOFF=OFF -DHPX_WITH_APEX=TRUE -DAPEX_WITH_ACTIVEHARMONY=TRUE -DACTIVEHARMONY_ROOT=/usr/local/packages/activeharmony/4.6.0-gcc -DAPEX_WITH_OTF2=TRUE -DOTF2_ROOT=/usr/local/packages/otf2-2.1 -DAPEX_WITH_PAPI=TRUE -DPAPI_ROOT=/usr/local/packages/papi/papi-knl/5.5.0/ -DHPX_WITH_APEX_NO_UPDATE=FALSE -DHPX_WITH_APEX_TAG=develop -DHPX_WITH_MAX_CPU_COUNT=72 /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx
  1. Make

## Specifications

... Please describe your environment

  - HPX Version: stable
  - Platform (compiler, OS): Linux, x86_64, gcc 7.1

For the records, here is the relevant error messgae:
```
[ 30%] Building CXX object src/CMakeFiles/hpx.dir/runtime/components/memory.cpp.o
In file included from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/pack_traversal.hpp:9:0,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/unwrap_impl.hpp:15,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/unwrap.hpp:10,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/parallel/executors/parallel_executor.hpp:33,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/dataflow.hpp:46,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/parallel/executors/execution.hpp:18,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/local/packaged_continuation.hpp:23,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/future.hpp:1715,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/performance_counters/counters.hpp:11,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/src/runtime/components/binpacking_distribution_policy.cpp:6:
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp: In instantiation of ‘auto hpx::util::detail::container_remapping::rebind_container(const Container&) [with NewType = long unsigned int; Container = std::vector<hpx::lcos::future<long unsigned int> >]’:
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:490:67:   required by substitution of ‘template<class M, class T> decltype (rebind_container<hpx::util::detail::container_remapping::mapped_type_from_t<T, M> >(container)) hpx::util::detail::container_remapping::remap_container(hpx::util::detail::container_remapping::container_mapping_tag<false, false>, M&&, T&&) [with M = hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_traversor; T = std::vector<hpx::lcos::future<long unsigned int> >&]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:558:44:   required by substitution of ‘template<class T, class M> decltype (hpx::util::detail::container_remapping::remap_container(hpx::util::detail::container_remapping::container_mapping_tag_of_t<T, M>{}, forward<M>(mapper), forward<T>(container))) hpx::util::detail::container_remapping::remap(hpx::util::detail::strategy_remap_tag, T&&, M&&, typename std::enable_if<hpx::traits::is_invocable<typename M::traversor_type, typename std::conditional<std::is_rvalue_reference<_Tp&&>::value, decltype (std::move((* declval<Container>().begin()))), decltype (* declval<Container>().begin())>::type>::value>::type*) [with T = std::vector<hpx::lcos::future<long unsigned int> >&; M = hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_traversor]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:837:55:   required by substitution of ‘template<class T> decltype (hpx::util::detail::container_remapping::remap(hpx::util::detail::strategy_remap_tag{}, forward<T>(container), declval<hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_traversor>())) hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_match<T>(hpx::util::detail::container_category_tag<true, false>, T&&) [with T = std::vector<hpx::lcos::future<long unsigned int> >&]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:870:18:   required by substitution of ‘template<class T> decltype (declval<hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> > >().hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_match(declval<hpx::util::detail::container_category_of_t<typename hpx::util::decay_unwrap<T>::type> >(), declval<T>())) hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_traverse<T>(hpx::util::detail::strategy_remap_tag, T&&) [with T = std::vector<hpx::lcos::future<long unsigned int> >&]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:893:54:   required by substitution of ‘template<class T> decltype (hpx::util::detail::spreading::unpack_or_void(declval<hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> > >().hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::try_traverse<T>(hpx::util::detail::strategy_remap_tag{}, declval<T>()))) hpx::util::detail::mapping_helper<hpx::util::detail::strategy_remap_tag, hpx::util::detail::future_unwrap_until_depth<1> >::init_traverse<T>(hpx::util::detail::strategy_remap_tag, T&&) [with T = std::vector<hpx::lcos::future<long unsigned int> >&]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:942:14:   required by substitution of ‘template<class Strategy, class Mapper, class ... T> decltype (declval<hpx::util::detail::mapping_helper<Strategy, typename std::decay<_Arg>::type> >().init_traverse(strategy, (forward<T>)(hpx::util::detail::apply_pack_transform::pack)...)) hpx::util::detail::apply_pack_transform(Strategy, Mapper&&, T&& ...) [with Strategy = hpx::util::detail::strategy_remap_tag; Mapper = hpx::util::detail::future_unwrap_until_depth<1>; T = {std::vector<hpx::lcos::future<long unsigned int>, std::allocator<hpx::lcos::future<long unsigned int> > >&}]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/pack_traversal.hpp:59:49:   required by substitution of ‘template<class Mapper, class ... T> decltype (hpx::util::detail::apply_pack_transform(hpx::util::detail::strategy_remap_tag{}, forward<Mapper>(mapper), (forward<T>)(hpx::util::map_pack::pack)...)) hpx::util::map_pack(Mapper&&, T&& ...) [with Mapper = hpx::util::detail::future_unwrap_until_depth<1>; T = {std::vector<hpx::lcos::future<long unsigned int>, std::allocator<hpx::lcos::future<long unsigned int> > >&}]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/unwrap_impl.hpp:129:33:   required by substitution of ‘template<long unsigned int Depth, class ... Args> decltype (hpx::util::map_pack(hpx::util::detail::future_unwrap_until_depth<Depth>{}, (forward<Args>)(hpx::util::detail::unwrap_depth_impl::args)...)) hpx::util::detail::unwrap_depth_impl(Args&& ...) [with long unsigned int Depth = 1; Args = {std::vector<hpx::lcos::future<long unsigned int>, std::allocator<hpx::lcos::future<long unsigned int> > >&}]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/unwrap.hpp:66:50:   required by substitution of ‘template<class ... Args> decltype (unwrap_depth_impl<1>((forward<Args>)(hpx::util::unwrap::args)...)) hpx::util::unwrap(Args&& ...) [with Args = {std::vector<hpx::lcos::future<long unsigned int>, std::allocator<hpx::lcos::future<long unsigned int> > >&}]’
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/src/runtime/components/binpacking_distribution_policy.cpp:81:52:   required from here
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:355:46: error: ambiguous template instantiation for ‘struct hpx::traits::pack_traversal_rebind_container<long unsigned int, std::vector<hpx::lcos::future<long unsigned int> >, void>’
                 return traits::pack_traversal_rebind_container<
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                     NewType, Container>::call(container);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
In file included from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/pack_traversal_impl.hpp:12:0,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/pack_traversal.hpp:9,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/detail/unwrap_impl.hpp:15,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/util/unwrap.hpp:10,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/parallel/executors/parallel_executor.hpp:33,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/dataflow.hpp:46,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/parallel/executors/execution.hpp:18,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/local/packaged_continuation.hpp:23,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/lcos/future.hpp:1715,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/performance_counters/counters.hpp:11,
                 from /var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/src/runtime/components/binpacking_distribution_policy.cpp:6:
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/traits/pack_traversal_rebind_container.hpp:21:12: note: candidates are: template<class NewType, template<class> class Base, class OldType> struct hpx::traits::pack_traversal_rebind_container<NewType, Base<OldType> > [with NewType = long unsigned int; Base = std::vector; OldType = hpx::lcos::future<long unsigned int>]
     struct pack_traversal_rebind_container<NewType, Base<OldType>>
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/buildbot/slaves/phylanx/x86_64-gcc7-debug/build/tools/buildbot/src/hpx/hpx/traits/pack_traversal_rebind_container.hpp:34:12: note:                 template<class NewType, template<class, class> class Base, class OldType, class OldAllocator> struct hpx::traits::pack_traversal_rebind_container<NewType, Base<OldType, OldAllocator>, typename std::enable_if<std::uses_allocator<Base<OldType, OldAllocator>, OldAllocator>::value>::type> [with NewType = long unsigned int; Base = std::vector; OldType = hpx::lcos::future<long unsigned int>; OldAllocator = std::allocator<hpx::lcos::future<long unsigned int> >]
     struct pack_traversal_rebind_container<NewType, Base<OldType, OldAllocator>,
            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         typename std::enable_if<std::uses_allocator<
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                 Base<OldType, OldAllocator>, OldAllocator>::value
                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             >::type>
             ~~~~~~~~
```

--- a/hpx/traits/pack_traversal_rebind_container.hpp
+++ b/hpx/traits/pack_traversal_rebind_container.hpp
@@ -8,45 +8,98 @@
 #define HPX_TRAITS_PACK_TRAVERSAL_REBIND_CONTAINER_JUN_10_2019_1020AM
 
 #include <memory>
+#include <list>
 #include <type_traits>
+#include <vector>
 
 namespace hpx { namespace traits
 {
-    ///////////////////////////////////////////////////////////////////////////
-    template <typename NewType, typename Base, typename Enable = void>
-    struct pack_traversal_rebind_container;
-
-    /// Specialization for a container with a single type T (no allocator support)
-    template <typename NewType, template <class> class Base, typename OldType>
-    struct pack_traversal_rebind_container<NewType, Base<OldType>>
+    ////////////////////////////////////////////////////////////////////////////
+    namespace detail
     {
-        static Base<NewType> call(Base<OldType> const& /*container*/)
-        {
-            return Base<NewType>();
-        }
-    };
+        ////////////////////////////////////////////////////////////////////////
+        template <typename NewType, typename OldType, typename Enable = void>
+        struct pack_traversal_rebind_container;
 
-    /// Specialization for a container with a single type T and
-    /// a particular allocator, which is preserved across the remap.
-    /// -> We remap the allocator through std::allocator_traits.
-    template <typename NewType, template <class, class> class Base,
-        typename OldType,typename OldAllocator>
-    struct pack_traversal_rebind_container<NewType, Base<OldType, OldAllocator>,
-        typename std::enable_if<std::uses_allocator<
-                Base<OldType, OldAllocator>, OldAllocator>::value
-            >::type>
+        // Specialization for a container with a single type T (no allocator
+        // support)
+        template <typename NewType, template <class> class Base,
+            typename OldType>
+        struct pack_traversal_rebind_container<NewType, Base<OldType>>
+        {
+            static Base<NewType> call(Base<OldType> const& /*container*/)
+            {
+                return Base<NewType>();
+            }
+        };
+
+        // Specialization for a container with a single type T and
+        // a particular allocator, which is preserved across the remap.
+        // -> We remap the allocator through std::allocator_traits.
+        template <typename NewType, template <class, class> class Base,
+            typename OldType, typename OldAllocator>
+        struct pack_traversal_rebind_container<NewType,
+            Base<OldType, OldAllocator>,
+            typename std::enable_if<std::uses_allocator<
+                Base<OldType, OldAllocator>, OldAllocator>::value>::type>
+        {
+            using NewAllocator = typename std::allocator_traits<OldAllocator>::
+                template rebind_alloc<NewType>;
+
+            static Base<NewType, NewAllocator> call(
+                Base<OldType, OldAllocator> const& container)
+            {
+                // Create a new version of the allocator, that is capable of
+                // allocating the mapped type.
+                return Base<NewType, NewAllocator>(
+                    NewAllocator(container.get_allocator()));
+            }
+        };
+    }
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Implement a two-level specialization to avoid ambiguities between the
+    // specializations for standard containers below and the generic fallback
+    // solutions provided above
+
+    template <typename NewType, typename OldType, typename Enable = void>
+    struct pack_traversal_rebind_container
+      : detail::pack_traversal_rebind_container<NewType, OldType>
+    {};
+
+    // gcc reports an ambiguity for any standard container that has a defaulted
+    // allocator template argument as it believes both specializations above are
+    // viable. This works around by explicitly specializing the trait.
+    template <typename NewType, typename OldType, typename OldAllocator>
+    struct pack_traversal_rebind_container<
+        NewType, std::vector<OldType, OldAllocator>>
     {
         using NewAllocator = typename std::allocator_traits<OldAllocator>::
             template rebind_alloc<NewType>;
 
-        // Check whether the second argument of the container was
-        // the used allocator.
-        static Base<NewType, NewAllocator> call(
-            Base<OldType, OldAllocator> const& container)
+        static std::vector<NewType, NewAllocator> call(
+            std::vector<OldType, OldAllocator> const& container)
         {
-            // Create a new version of the allocator, that is capable of
-            // allocating the mapped type.
-            return Base<NewType, NewAllocator>(
+            // Create a new version of the container using the new allocator
+            // that is capable of allocating the mapped type.
+            return std::vector<NewType, NewAllocator>(
+                NewAllocator(container.get_allocator()));
+        }
+    };
+
+    template <typename NewType, typename OldType, typename OldAllocator>
+    struct pack_traversal_rebind_container<
+        NewType, std::list<OldType, OldAllocator>>
+    {
+        using NewAllocator = typename std::allocator_traits<OldAllocator>::
+            template rebind_alloc<NewType>;
+
+        static std::list<NewType, NewAllocator> call(
+            std::list<OldType, OldAllocator> const& container)
+        {
+            // Create a new version of the container using the new allocator
+            // that is capable of allocating the mapped type.
+            return std::list<NewType, NewAllocator>(
                 NewAllocator(container.get_allocator()));
         }
     };

